src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -4255,23 +4184,33 @@ def _array_syntax_to_indexed(self, parent, loop_vars):
         :raises NotImplementedError: if array sections of differing ranks are
                                      found.
         '''
-        assigns = parent.walk(Assignment)
-        # Check that the LHS of any assignment uses array notation.
-        # Note that this will prevent any WHERE blocks that contain scalar
-        # assignments from being handled but is a necessary limitation until
-        # #717 is done and we interrogate the type of each symbol.
-        for assign in assigns:
-            _ = self._array_notation_rank(assign.lhs)
-
-        # TODO #717 if the supplied code accidentally omits array
-        # notation for an array reference on the RHS then we will
-        # identify it as a scalar and the code produced from the
-        # PSyIR (using e.g. the Fortran backend) will not
-        # compile. We need to implement robust identification of the
-        # types of all symbols in the PSyIR fragment.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.transformations import (
+                Reference2ArrayRangeTrans, TransformationError)
+        # Convert References to arrays to use the array range notation unless
+        # they have an IntrinsicCall parent.
+        for ref in parent.walk(Reference):
+            if isinstance(ref.symbol.interface, ImportInterface):
+                raise NotImplementedError(
+                        "PSyclone doesn't yet support reference to imported "
+                        "symbols inside WHERE clauses.")
+            intrinsic_ancestor = ref.ancestor(IntrinsicCall)
```

Comment:
Not only Intrinsics, this will also need to be checked for any Call. If its elemental, its arguments must have array notation (for the next step to add the index), if it's not elemental, it shouldn't as it takes the whole array as argument.

For calls there is the additional case of when we don't know if it's elemental or not. In this case it should raise NotImplementedError.

Can you add tests for the 3 cases?

Reply:
Do we need explicit tests for Call since we it will use the same code as IntrinsicCall for most of this, only the NotImplementedError case I should have a separate test.

Reply:
Yes, if they use the same code you can only test the new NotImplementedError case

Reply:
This should be covered then.